warehouse/integrations/redshift/redshift.go

```diff
@@ -969,6 +963,87 @@ func (rs *Redshift) connect(ctx context.Context) (*sqlmiddleware.DB, error) {
 	return middleware, nil
 }
 
+func (rs *Redshift) useIAMForAuth() bool {
+	return warehouseutils.ReadAsBool(configKeyUseIAMForAuth, rs.Warehouse.Destination.Config)
+}
+
+func (rs *Redshift) connectUsingIAMRole() (*sql.DB, error) {
+	var (
+		database          = warehouseutils.GetConfigValue(configKeyDatabase, rs.Warehouse)
+		user              = warehouseutils.GetConfigValue(configKeyUser, rs.Warehouse)
+		iamRoleARNForAuth = warehouseutils.GetConfigValue(configKeyIAMRoleARNForAuth, rs.Warehouse)
+		clusterID         = warehouseutils.GetConfigValue(configKeyClusterID, rs.Warehouse)
+		clusterRegion     = warehouseutils.GetConfigValue(configKeyClusterRegion, rs.Warehouse)
+		timeout           = rs.connectTimeout
+	)
+
+	data := sqlconnectconfig.RedshiftData{
+		ClusterIdentifier: clusterID,
+		Database:          database,
+		User:              user,
+		Region:            clusterRegion,
+		RoleARN:           iamRoleARNForAuth,
+		ExternalID:        rs.Warehouse.WorkspaceID,
+		RoleARNExpiry:     time.Hour,
+		Timeout:           timeout,
+	}
+
+	credentialsJSON, err := data.MarshalJSON()
+	if err != nil {
+		return nil, fmt.Errorf("connectUsingIAMRole: marshalling redshift credentials: %w", err)
+	}
+
+	sqlConnectDB, err := sqlconnect.NewDB("redshift", credentialsJSON)
+	if err != nil {
+		return nil, fmt.Errorf("connectUsingIAMRole: creating redshift connection: %w", err)
+	}
+	return sqlConnectDB.SqlDB(), nil
+}
+
+func (rs *Redshift) connectUsingPassword() (*sql.DB, error) {
+	var (
+		host       = warehouseutils.GetConfigValue(configKeyHost, rs.Warehouse)
+		port       = warehouseutils.GetConfigValue(configKeyPort, rs.Warehouse)
+		database   = warehouseutils.GetConfigValue(configKeyDatabase, rs.Warehouse)
+		user       = warehouseutils.GetConfigValue(configKeyUser, rs.Warehouse)
+		password   = warehouseutils.GetConfigValue(configKeyPassword, rs.Warehouse)
+		timeout    = rs.connectTimeout
+		tunnelInfo = tunnelling.ExtractTunnelInfoFromDestinationConfig(rs.Warehouse.Destination.Config)
+	)
+
+	var (
+		err error
+		db  *sql.DB
+	)
+
+	dsn := url.URL{
+		Scheme: "postgres",
+		User:   url.UserPassword(user, password),
+		Host:   fmt.Sprintf("%s:%s", host, port),
+		Path:   database,
+	}
+
+	params := url.Values{}
+	params.Add("sslmode", "require")
+
+	if timeout > 0 {
+		params.Add("connect_timeout", fmt.Sprintf("%d", timeout/time.Second))
```

Comment:
if timeout < time.Second, we will set `connect_timeout=0` .

Does connect_timeout supports decimals ?
What the behaviour if it is zero? Does it mean no timeout ?

It might be unexpected if someone configures 500ms, and there is actually no connection timeout.

We can consider ceiling the division or throw an error if less than a second is provided

Reply:
It's better to throw error is `timeout < time.Second`. Making the changes.